Feature/kafka 2.12 1.0.0

vegiapparao opened the pull request:
This PR has the changes to get Apache Flume working with the latest version if Kafka client.  We had issues getting one of the authentication types working with older version of Kafka client.  Its working with Kafka_2.12_1.0.0.

Few tests are failing (most of the times they are related to VM crash/time out)  in flume-kafka-channel project, I will be fixing them, but I am creating this PR for an early feedback.

rgoers committed:
FLUME-2050 Upgrade to Log4j 2

rgoers committed:
FLUME-2050 remove spurious files

rgoers committed:
FLUME-2050 - modify pattern layout so NDC is ignored if it has no data

rgoers committed:
Update configuration to match log4j 1.x

rgoers committed:
Remove more references to log4j 1

Commit:
FLUME-3195. Split the KafkaChannelTest to avoid timeouts

KafkaChannelTest had quite a few test methods so during the build sometimes it caused a timeout. Refactoring it to smaller test classes decrease the chance of timeouts.

This closes #183

Reviewers: Miklos Csanady, Ferenc Szabo

(Viktor Somogyi via Ferenc Szabo)

Commit:
FLUME-3158: Upgrade surefire version and config

Set surefire version to the latest and configure it
to rerun failed tests.
This way the flaky tests will not break the build as often.

This closes #182

Reviewers: Ferenc Szabo

(Miklos Csanady via Ferenc Szabo)

rgoers committed:
Upgrade to Log4j 2.10.0

rgoers committed:
Merge branch 'trunk' into flume-2050

Commit:
Changes made to get Apache Flume working with the latest version of kafka client 2_12.1.0.0

Commit:
Fixed checkstyle errors

tmgstevens commented on flume-ng-sinks/flume-ng-kafka-sink/src/test/java/org/apache/flume/sink/kafka/util/TestUtil.java:
```diff
@@ -57,6 +57,7 @@ private void init() {
     // get the localhost.
     try {
       hostname = InetAddress.getLocalHost().getHostName();
+      hostname = "localhost";
```
Please remove this. This is probably due to an incorrect /etc/hosts file in your test environment

tmgstevens commented on flume-ng-sources/flume-kafka-source/pom.xml:
```diff
@@ -83,9 +83,8 @@
       <groupId>org.apache.kafka</groupId>
-      <artifactId>kafka_2.10</artifactId>
-      <classifier>test</classifier>
-      <scope>test</scope>
```
test scope should remain for this - we only need full Kafka for the tests, right?

tmgstevens commented on flume-ng-sources/flume-kafka-source/src/test/java/org/apache/flume/source/kafka/TestKafkaSource.java:
```diff
@@ -46,6 +46,7 @@
 import org.apache.kafka.common.security.JaasUtils;
 import org.junit.After;
 import org.junit.Before;
+import org.junit.Ignore;
```
What is this for?

tmgstevens commented on flume-ng-sources/flume-kafka-source/src/test/java/org/apache/flume/source/kafka/TestKafkaSource.java:
```diff
@@ -20,7 +20,7 @@
 import com.google.common.collect.Lists;
 import junit.framework.Assert;
-import kafka.common.TopicExistsException;
+import org.apache.kafka.common.errors.TopicExistsException;
```
What is this for?

tmgstevens commented on flume-ng-sources/flume-kafka-source/src/main/java/org/apache/flume/source/kafka/KafkaSource.java:
```diff
@@ -57,16 +46,31 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Pattern;
+
+import static org.apache.flume.source.kafka.KafkaSourceConstants.DEFAULT_GROUP_ID;
```
What are these imports for? Please review.

vegiapparao replied:
Sorry its my IntelliJ settings, we have a policy not to use wildcard imports.  I can revert it back if it is not acceptable.  Please advise.

tmgstevens replied:
I think our style guide allows wildcard imports for static fields. Please try to ensure that non-stylistic changes get brought in when submitting a patch.

vegiapparao replied:
Import style changes have been reverted.

tmgstevens commented on flume-ng-sdk/src/test/java/org/apache/flume/event/TestEventBuilder.java:
```diff
@@ -19,14 +19,14 @@
 
+import java.util.HashMap;
+import java.util.Map;
+
```
These changes are redundant

tmgstevens commented on flume-ng-channels/flume-kafka-channel/src/test/resources/kafka-server.properties:
```diff
@@ -18,20 +18,24 @@
 # java.net.InetAddress.getCanonicalHostName().
 #advertised.host.name=<hostname routable by clients>
+advertised.listeners=PLAINTEXT://localhost:9092
+advertised.host.name = localhost
```
We shouldn't need these changes

Commit:
Address review comments

Commit:
Removed redundant properties from the kafka-server.properties

szaboferee reviewed (changes requested):
Thanks for the contribution!
What is the JIRA ticket number for this change? We have to track it for the releases.

Will it be backward compatible with the kafka cluster versions that the current client version is compatible with?

vegiapparao commented:
Bidirectional support came in with version 0.10.2 here is the JIRA tracking references
(https://issues.apache.org/jira/browse/KAFKA-3304) & (https://issues.apache.org/jira/browse/KAFKA-4462)


With this PR, it updates the client to 1.0.0.  We have tested it against a Kafka broker version < 1.0.0
Compatibility matrix can be found here: https://cwiki.apache.org/confluence/display/KAFKA/Compatibility+Matrix

vegiapparao commented:
@tmgstevens, @szaboferee 
Can we link this to this issue? https://issues.apache.org/jira/browse/FLUME-3197
How do I get the issue on to the board and assign it to me?

Thanks

tmgstevens commented:
@vegiapparao I can't assign it to you as I think you need to be added to the project first.

@szaboferee This one does give us a slight compatibility headache. We currently support Kafka 0.9.0 and this would give us 0.10.0+ (including 1.0.0+). The newer versions of Kafka have version negotiation on the wire protocol (https://cwiki.apache.org/confluence/display/KAFKA/KIP-35+-+Retrieving+protocol+version), so this would, presumably, be the last time we'd have to break compatibility.

We also broke compatibility in https://issues.apache.org/jira/browse/FLUME-2820 in the same way.

tmgstevens commented on flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:
```diff
@@ -76,8 +77,32 @@
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 
-import static org.apache.flume.channel.kafka.KafkaChannelConfiguration.*;
```
I believe our style guide allows wildcard imports for static fields

vegiapparao replied:
Reverted the change

tmgstevens replied:
Is this still outstanding?

tmgstevens commented on flume-ng-channels/flume-kafka-channel/src/main/java/org/apache/flume/channel/kafka/KafkaChannel.java:
```diff
@@ -296,8 +321,9 @@ private synchronized ConsumerAndRecords createConsumerAndRecords() {
       KafkaConsumer<String, byte[]> consumer = new KafkaConsumer<String, byte[]>(consumerProps);
       ConsumerAndRecords car = new ConsumerAndRecords(consumer, channelUUID);
       logger.info("Created new consumer to connect to Kafka");
-      car.consumer.subscribe(Arrays.asList(topic.get()),
```
I'm not sure why you've changed this.

tmgstevens replied:
Why are we adding in a dangling ChannelRebalanceListener. Can we be sure that this doesn't introduce bugs?

vegiapparao replied:
@tmgstevens, references to ChannelRebalanceListener has been removed.  There was a compilation error when I first started making changes and hence I added `ChannelRebalanceListener`.

tmgstevens replied:
Still confused here. It looks like you've removed the `ChannelRebalanceListener` altogether. I don't think we need to do this. Please add this back in.

tmgstevens commented on flume-ng-channels/flume-kafka-channel/src/test/java/org/apache/flume/channel/kafka/TestBasicFunctionality.java:
```diff
@@ -49,6 +49,7 @@ public void testProps() throws Exception {
     Context context = new Context();
     context.put("kafka.producer.some-parameter", "1");
     context.put("kafka.consumer.another-parameter", "1");
+    context.put("offsets.topic.replication.factor", "1");
```
What does this change bring? Did we need this to get the test working?

vegiapparao replied:
Removed it now.  Initially, the tests failed with an error and an online search suggested to have this property in kafka-server.properties.

vegiapparao replied:
The tests intermittently fail with this error.
`Number of alive brokers '1' does not meet the required replication factor '3' for the offsets topic (configured via 'offsets.topic.replication.factor'). This error can be ignored if the cluster is starting up and not all brokers are up yet`.  The tests would pass consistently by having the property `offsets.topic.replication.factor = 1` in `kafka-server.properties`.

tmgstevens commented on flume-ng-channels/flume-kafka-channel/src/test/java/org/apache/flume/channel/kafka/TestKafkaChannelBase.java:
```diff
@@ -120,6 +122,7 @@ Context prepareDefaultContext(boolean parseAsFlume) {
     Context context = new Context();
     context.put(BOOTSTRAP_SERVERS_CONFIG, testUtil.getKafkaServerUrl());
     context.put(PARSE_AS_FLUME_EVENT, String.valueOf(parseAsFlume));
+    context.put("offsets.topic.replication.factor", "1");
```
What does this change bring? Did we need this to get the test working?

vegiapparao replied:
Removed it now.

tmgstevens commented on flume-ng-channels/flume-kafka-channel/src/test/java/org/apache/flume/channel/kafka/TestOffsetsAndMigration.java:
```diff
@@ -121,6 +121,7 @@ private void doTestMigrateZookeeperOffsets(boolean hasZookeeperOffsets, boolean
     Context context = prepareDefaultContext(false);
     context.put(ZOOKEEPER_CONNECT_FLUME_KEY, testUtil.getZkUrl());
     context.put(GROUP_ID_FLUME, group);
+    context.put("offsets.topic.replication.factor", "1");
```
What does this change bring? Did we need this to get the test working?

vegiapparao replied:
Removed it now.

tmgstevens commented on flume-ng-sources/flume-kafka-source/src/main/java/org/apache/flume/source/kafka/KafkaSource.java:
```diff
@@ -16,21 +16,10 @@
  */
 package org.apache.flume.source.kafka;
 
-import java.io.ByteArrayInputStream;
```
Please tidy this commit to avoid reformatting changes (for example reordering of imports)

tmgstevens commented on flume-ng-sources/flume-kafka-source/src/main/java/org/apache/flume/source/kafka/KafkaSource.java:
```diff
@@ -464,12 +468,17 @@ private String lookupBootstrap(String zookeeperConnect, SecurityProtocol securit
+      for (Broker broker : brokers) {
+        List<EndPoint> endPointList = seqAsJavaList(broker.endPoints());
+        for (EndPoint endPoint : endPointList) {
+          connections.add(endPoint.host() + ":" + endPoint.port());
```
I'm intrigued by this change - is endPoint.connectionString no longer supplied? It seems the old version would have been better.

vegiapparao replied:
`endPoint.connectingString` does exist in the new library, but the newer one returns additional text.  This is evident from the test `TestKafkaSource#testBootstrapLookup`.  The test fails with the following output.  

Expected :localhost:65069
Actual   :PLAINTEXT://localhost:65069

tmgstevens replied:
Is there something we need to be aware of here when having non-plaintext listeners?

vegiapparao replied:
Sorry @tmgstevens I have no idea about that.  The test fails with this error if I change it to `endPoint.connectingString`

[ERROR] Tests run: 36, Failures: 11, Errors: 0, Skipped: 0, Time elapsed: 311.043 s <<< FAILURE! - in org.apache.flume.source.kafka.TestKafkaSource
[ERROR] testBootstrapLookup(org.apache.flume.source.kafka.TestKafkaSource)  Time elapsed: 4.101 s  <<< FAILURE!
junit.framework.ComparisonFailure: expected:<[]localhost:51213> but was:<[PLAINTEXT://]localhost:51213>
	at org.apache.flume.source.kafka.TestKafkaSource.testBootstrapLookup(TestKafkaSource.java:607)

tmgstevens replied:
I think this looks okay, but I'd like to test it in a secure environment.

tmgstevens replied:
On reflection I'm not happy with this. I haven't had the chance to test it in a secure environment, but I think the test is wrong. It is going to be important to prefix with the type of listener that is being used. For the test we can prefix PLAINTEXT, but if it's SASL_SSL, or any other type of listener then I think we'd need to provide that back from there.

vegiapparao replied:
@tmgstevens  apologies for the delay. I am happy to change the test in that case, but I have no idea about testing other types / neither do I have the environment to test the rest of the types.  Any advise?

tmgstevens replied:
Please could you try this code:
```
private String lookupBootstrap(String zookeeperConnect, SecurityProtocol securityProtocol) {
    ZkUtils zkUtils = ZkUtils.apply(zookeeperConnect, ZK_SESSION_TIMEOUT, ZK_CONNECTION_TIMEOUT,
        JaasUtils.isZkSecurityEnabled());
    try {
      Seq<Broker> allBrokersInCluster = zkUtils.getAllBrokersInCluster();
      List<Broker> brokerList = JavaConverters.seqAsJavaListConverter(zkUtils.getAllBrokersInCluster()).asJava();
      List<BrokerEndPoint> endPoints = brokerList.stream()
              .map(broker -> broker.getBrokerEndPoint(
                  ListenerName.forSecurityProtocol(securityProtocol))
              )
              .collect(Collectors.toList());
      List<String> connections = new ArrayList<>();
      for (BrokerEndPoint endPoint : endPoints) {
        connections.add(endPoint.connectionString());
      }
      return StringUtils.join(connections, ',');
    } finally {
      zkUtils.close();
    }
  }
```

tmgstevens commented on flume-ng-sources/flume-kafka-source/src/test/java/org/apache/flume/source/kafka/KafkaSourceEmbeddedKafka.java:
```diff
@@ -67,6 +68,7 @@ public KafkaSourceEmbeddedKafka(Properties properties) {
     Properties props = new Properties();
     props.put("zookeeper.connect",zookeeper.getConnectString());
     props.put("broker.id","1");
+    props.put("offsets.topic.replication.factor", "1");
```
Please remove if not req'd

vegiapparao replied:
Removed

tmgstevens replied:
I think we still need this, right?

tmgstevens commented on flume-ng-sources/flume-kafka-source/src/test/java/org/apache/flume/source/kafka/TestKafkaSource.java:
```diff
@@ -69,14 +69,14 @@
 import static org.apache.flume.source.kafka.KafkaSourceConstants.BATCH_SIZE;
 import static org.apache.flume.source.kafka.KafkaSourceConstants.BOOTSTRAP_SERVERS;
 import static org.apache.flume.source.kafka.KafkaSourceConstants.DEFAULT_AUTO_COMMIT;
+import static org.apache.flume.source.kafka.KafkaSourceConstants.DEFAULT_TOPIC_HEADER;
```
Please avoid reordering imports

tmgstevens commented on flume-shared/flume-shared-kafka-test/src/main/java/org/apache/flume/shared/kafka/test/KafkaPartitionTestUtil.java:
```diff
@@ -199,7 +200,9 @@ public static void checkResultsAgainstSkew(PartitionTestScenario scenario,
 
       TopicPartition partition = new TopicPartition(topic, i);
 
-      consumer.assign(Arrays.asList(partition));
```
What does this change bring?

vegiapparao replied:
`KafkaConsume.assign` signature has been changed in 0.10.0.0
0.9.0.1 => `public void assign(java.util.List<org.apache.kafka.common.TopicPartition>);`
0.10.0.0 => `public void assign(java.util.Collection<org.apache.kafka.common.TopicPartition>);`

tmgstevens replied:
+1

Commit:
Implemented PR comments
* Removed 'offsets.topic.replication.factor' from the source files
* Reverted import style changes

vegiapparao commented:
@tmgstevens - I managed to get all the tests working locally by modifying /etc/hosts.  There is only one test failing as of now  `TestSyslogUtils#TestRfc3164Dates`.  Are you aware of any issues with that test?

tmgstevens commented:
I wrote that test and yes, it's failing for me. Probably because it's December (which is a bug). Would you mind raising a JIRA for that and I'll pick it up?

vegiapparao commented:
@tmgstevens I was just about to raise the JIRA and submit a patch, but found this issue here
https://issues.apache.org/jira/browse/FLUME-3201, may be related, but sounds like an issue with the code in the TestSyslogUtils#TestRfc3164Dates:201 itself.  Should this line 
`SimpleDateFormat sdf = new SimpleDateFormat("MMM  d hh:MM:ss");` 
be changed to
`SimpleDateFormat sdf = new SimpleDateFormat("MMM dd HH:mm:ss");`

I ran it locally with the change and the tests are passing.  Happy to submit the patch if you are ok.  Let me know.  Thanks

ssvinarchuk commented:
As I see all tests are passed now. So could you say when this PR will merge?
Thanks

tmgstevens commented:
@ssvinarchuk @vegiapparao I'd like to see this committed in time for Flume 1.9, but there's a number of my review comments outstanding, or that haven't been reflected in the latest commits. I'm particularly worried about the change to the KafkaChannel with the dangling ChannelRebalanceListener.

@vegiapparao will you have a chance to make these change?

vegiapparao commented:
@tmgstevens yes, I will address your comments very soon.  Sorry about the delay.

ssvinarchuk commented:
@vegiapparao Do you have time to finish all needed changes to merge this PR?

vegiapparao commented:
hi @ssvinarchuk, @tmgstevens - Apologies for the delay.  Stuck with an internal release.  I will work on the changes today, will push them by end of today or early tomorrow.  Sorry again.

vegiapparao commented:
@ssvinarchuk, @tmgstevens having few test failures after implementing PR comments. Hoping to fix and push the changes soon. Apologies again for the delay. Thank you.

szaboferee commented:
@vegiapparao there was a force push to the trunk whit some git history fix, to be able to merge your change, you need to rebase it on top of the new trunk.

Commit:
Removed references to ChannelRebalanceListener from KafkaChannel as part of PR comment.  Its redundant.
There were couple of conflicts when rebased from upstream trunk.

Commit:
Removed unused import

Commit:
Remove references to ConsumerRebalanceListener

vegiapparao commented:
@tmgstevens, all tests pass locally.  But I had to make couple of test configs for the tests to run.  Add `offsets.topic.replication.factor=1`, change the value of `broker.id` from `0` to `-1` in `kafka-server.properties` under `flume-kafka-channel` and `flume-ng-kafka-sink` projects.  With out these changes, the tests fail with a timeout error.

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.20.1:test (default-test) on project flume-ng-kafka-sink: There was a timeout or other error in the fork -> [Help 1]

The below error logged in the test reports:
2018-02-16 02:44:21,958 (kafka-request-handler-0) [ERROR - kafka.utils.Logging.error(Logging.scala:101)] [KafkaApi-0] Number of alive brokers '1' does not meet the required replication factor '3' for the offsets topic (configured via 'offsets.topic.replication.factor'). This error can be ignored if the cluster is starting up and not all brokers are up yet.

Is it a known issue?  I haven't committed/pushed config changes.

Apologies for the multiple commits with minor changes.

tmgstevens commented:
This would make sense - clearly in the unit tests we're only running a single broker so the offsets topic would only ever have one replica.

vegiapparao commented:
@tmgstevens, should I commit the test config changes or leave them as is?  How would the tests run for you?

ckdarby commented:
@tmgstevens This branch introduces support for flume to use a Kafka client >= 0.10 ?

There is no other way until this is merged?

tmgstevens commented:
@ckdarby This is correct. The Kafka API is unstable enough that it's not possible just to drop in a new version. That said, the current (0.9) client will work with 0.10 and 1.0 brokers as I understand. https://cwiki.apache.org/confluence/display/KAFKA/Compatibility+Matrix

jrpilat commented:
While the current (0.9) client will work with 0.10 and 1.0 brokers, there are still some caveats described [here](https://kafka.apache.org/documentation/#upgrade_10_performance_impact).

Essentially, for any topic of significant throughput, you must either:
- Use the older on-disk message format, and forego some of the features that come with the newer format, **or**
- Use the newer on-disk message format, in which case the newer broker will downconvert the messages on-the-fly to the older format.  If you do this, the broker cannot use zero-copy to write the message directly from disk to NIC, and consequently it is a huge performance impact.

vegiapparao commented:
@tmgstevens, still waiting for your advice on the test config changes, should I commit them or not.  I think I have addressed all your comments.  Thanks

tmgstevens commented:
Hi @vegiapparao - yes please push those changes to your branch.

I'll then take a final look and run through of the tests.

vegiapparao commented:
@tmgstevens I will push them and let you know by end of today (NZDT).

Commit:
This commit has the following changes
a. Re-instated wild card imports
b. Changed the broker id to -1 and set the offset replication factor to 1 in tests

vegiapparao commented:
@tmgstevens I pushed all the changes.

vegiapparao commented:
@tmgstevens did you get chance to look at the changes?

tmgstevens commented:
Sorry, I've been away - will try to get to it this week.

asfgit commented:
Can one of the admins verify this patch?

vegiapparao commented:
@tmgstevens, not sure if this PR is still valid.  Happy to work further if required.  Please advise.

tmgstevens commented:
@vegiapparao Two comments from me. I think we should try and get this in to Flume 1.9 though.

Question for you: If you change the pom.xml to refer to Kafka 2.0.0, does it still compile and pass tests? If so, I think we should do that.

tmgstevens commented:
@vegiapparao Looks like this has been superseded by #235 . I propose we close. Sorry for the delays.

vegiapparao commented:
ok, thank you @tmgstevens.  I will close for now, but happy to work on it if required.
